### PR TITLE
Everyone gets their own colour! (All the targets, anyway.)

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #7944

## Feature Details
Uses separately defined colours instead of the radar colours for the following things:
- The outline of the targeted ship in the target HUD.
- The target pointers around selected ships.
- The target pointers around minables within asteroid scanning range.
- The target pointers around the selected landable.

Also, landables now have more than just the friendly or hostile colour options for their pointers now:
- Dominated landables now have solid white pointers.
- Restricted landables now have the same colour as radar unfriendly, rather than the radar hostile colour, but have their own colour definition so it can be anything.

## UI Screenshots
~~Just go play the game.~~

## Usage Examples
N/A

## Testing Done
Played the game a little. Stuff seems to still have colours, which means the newly defined colours are being used (because it's certainly not using the radar colours for everything anymore.)

### Automated Tests Added
N/A

## Performance Impact
N/A, I hope.

### CODE REVIEW

File: source/Engine.cpp [hunk 1]
```diff
@@ -199,6 +199,87 @@ namespace {
 		}
 	}
 
+	const Color &GetTargetOutlineColor(int type)
+	{
+		const Color &player = *GameData::Colors().Get("ship target outline player");
```
Comment (quyykk) [inline source/Engine.cpp:204]: If those variables are only used once, why have them in the first place?

File: source/Engine.cpp [hunk 2]
```diff
@@ -449,6 +530,8 @@ void Engine::Wait()
 // Begin the next step of calculations.
 void Engine::Step(bool isActive)
 {
+	static const Color &playerShipSelectedPointerColor = *GameData::Colors().Get("ship target pointer player");
```
Comment (quyykk) [inline source/Engine.cpp:533]: You use this also only once, why the variable? Also the static is a big no-no.

File: source/MapPlanetCard.cpp [hunk 3]
```diff
@@ -44,10 +44,23 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
-		planet->CanLand() ? "Friendly" : "Restricted";
+	if(!hasSpaceport)
+		reputationLabel = "No Spaceport";
+	else
+	{
+		Planet::Friendliness friendliness = planet->GetFriendliness();
+		switch(friendliness)
+		{
+			case Planet::Friendliness::FRIENDLY:
+				reputationLabel = "Friendly";
```
Comment (quyykk) [inline source/MapPlanetCard.cpp:55]: You need breaks here, or else the cases will all fallthrough.

File: source/Engine.cpp [hunk 4]
```diff
@@ -199,6 +199,87 @@ namespace {
...
+		{
+			case Planet::Friendliness::FRIENDLY:
+				return friendly;
+			case Planet::Friendliness::RESTRICTED:
+				return restricted;
+			case Planet::Friendliness::HOSTILE:
+				return hostile;
+			case Planet::Friendliness::DOMINATED:
+				return dominated;
+		}
```
Comment (warp-core) [inline source/Engine.cpp:245]:
```suggestion
		}
		return restricted;
```

**Comment by Saugia** [comment]
~~Can we get some image examples?~~

**Comment by RisingLeaf** [comment]
demanding screenshots

File: source/Engine.cpp (hunk 1 above)
Comment (warp-core) [reply source/Engine.cpp:204]: Puts them all together, looks nicer.
I can move them into the if statement if you think that'd be better.

File: source/Engine.cpp (hunk 2 above)
Comment (warp-core) [reply source/Engine.cpp:534]: I wasn't sure how many times I'd use it at the time.
Shouldn't the static thing be fine, though?
~~Isn't everything in the namespace static?~~

File: source/MapPlanetCard.cpp (hunk 3 above)
Comment (warp-core) [reply source/MapPlanetCard.cpp:55]: oops

File: source/Engine.cpp (hunk 2 above)
Comment (quyykk) [reply source/Engine.cpp:534]: The issue is what you assign to the static: When in the future we'll change it so that plugins can be reloaded without closing the game, this value would still refer to the original one.

**Comment by warp-core** [comment]
> ~Can we get some image examples?~

~You want screenshots of the game looking the same? Do you not know what it looks like? Have you not played it???~

Restricted planet no longer hostile colours:
image

Dominated planet:
image

File: source/Engine.cpp (hunk 1 above)
Comment (quyykk) [reply source/Engine.cpp:204]: dewit

File: source/Engine.cpp [hunk 5]
```diff
@@ -199,6 +199,67 @@ namespace {
...
+		else if(type == Radar::BLINK)
+			return *GameData::Colors().Get("ship target pointer blink");
+		else
+			return *GameData::Colors().Get("ship target pointer inactive");
+	}
+
+	const Color &GetMinablePointerColor(bool selected)
+	{
+		const Color &targeted = *GameData::Colors().Get("minable target pointer selected");
+		const Color &untargeted = *GameData::Colors().Get("minable target pointer unselected");
```
Comment (quyykk) [inline source/Engine.cpp:257]: here too

File: data/interfaces.txt [hunk 6]
```diff
@@ -107,6 +107,34 @@ color "radar blink" 1. 1. 1. 0.
 color "radar viewport" 0. .3 0. 0.
 color "radar star" .6 .6 .6 0.
 
+# Colors used for the outline of the targeted ship.
+color "ship target outline player" .2 1. 0. 0.
+color "ship target outline friendly" .4 .6 1. 0.
+color "sihp target outline unfriendly" .8 .8 .3 0.
```
Comment (Hurleveur) [inline data/interfaces.txt:113]:
```suggestion
color "ship target outline unfriendly" .8 .8 .3 0.
```

File: source/Engine.cpp [hunk 7]
```diff
@@ -199,6 +199,65 @@ namespace {
...
+		else if(type == Radar::BLINK)
+			return *GameData::Colors().Get("ship target pointer blink");
+		else
+			return *GameData::Colors().Get("ship target pointer inactive");
+	}
+
+	const Color &GetMinablePointerColor(bool selected)
+	{
+		if(selected)
+			return *GameData::Colors().Get("minable target pointer selected");;
```
Comment (warp-core) [inline source/Engine.cpp:257]:
```suggestion
			return *GameData::Colors().Get("minable target pointer selected");
```

File: source/MapPlanetCard.cpp [hunk 8]
```diff
@@ -44,10 +44,26 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
+				break;
+			case Planet::Friendliness::RESTRICTED:
+				reputationLabel = "Restricted";
+				break;
+			case Planet::Friendliness::HOSTILE:
+				reputationLabel = "Hostile";
+				break;
+			case Planet::Friendliness::DOMINATED:
+				reputationLabel = "Dominated";
```
Comment (quyykk) [inline source/MapPlanetCard.cpp:64]: Missing break here.

File: source/MapPlanetCard.cpp [hunk 9]
```diff
@@ -44,10 +44,26 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
 
-	reputationLabel = !hasSpaceport ? "No Spaceport" :
-		GameData::GetPolitics().HasDominated(planet) ? "Dominated" :
-		planet->GetGovernment()->IsEnemy() ? "Hostile" :
-		planet->CanLand() ? "Friendly" : "Restricted";
+	if(!hasSpaceport)
+		reputationLabel = "No Spaceport";
+	else
+	{
+		Planet::Friendliness friendliness = planet->GetFriendliness();
```
Comment (quyykk) [inline source/MapPlanetCard.cpp:51]: You don't need this variable

File: source/Radar.cpp [hunk 10]
```diff
@@ -30,8 +30,8 @@ const int Radar::UNFRIENDLY = 2;
 const int Radar::HOSTILE = 3;
 const int Radar::INACTIVE = 4;
 const int Radar::SPECIAL = 5;
-const int Radar::ANOMALOUS = 6;
-const int Radar::BLINK = 7;
+const int Radar::BLINK = 6;
+const int Radar::ANOMALOUS = 7;
```
Comment (quyykk) [inline source/Radar.cpp:34]: Why are you switching the order?

Comment (Hurleveur) [reply source/Radar.cpp:34]: This got me confused as well

Comment (warp-core) [reply source/Radar.cpp:34]: I was doing something differently before, but I think I'm not doing that anymore.

File: source/Engine.h [hunk 11]
```diff
@@ -135,7 +135,7 @@ class Engine {
 		Point center;
 		Angle angle;
 		double radius;
-		int type;
+		const Color &color;
```
Comment (Hurleveur) [inline source/Engine.h:138]: Not sure why was this changed? I dont see this new variable being used but mb I missed it

File: source/MapPlanetCard.cpp [hunk 12]
```diff
@@ -44,10 +44,26 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
-		planet->GetGovernment()->IsEnemy() ? "Hostile" :
-		planet->CanLand() ? "Friendly" : "Restricted";
+	if(!hasSpaceport)
+		reputationLabel = "No Spaceport";
+	else
+	{
+		switch(planet->GetFriendliness())
+		{
+			case Planet::Friendliness::FRIENDLY:
+				reputationLabel = "Friendly";
```
Comment (Hurleveur) [inline source/MapPlanetCard.cpp:54]: Shouldnt we have a default case; Friendly?
Also this seems unrelated but its fine

**Comment by Hurleveur** [comment]
The code itself looks good these are just small notes

File: source/Engine.h (hunk 11 above)
Comment (warp-core) [reply source/Engine.h:138]: I changed this so each set of target pointers (this class, Engine::Target holds the parameters for the target pointers) can have it's own colour instead of just refering to one of the Radar types.
It is used in Engine::Draw() when drawing the target pointers where the shadar is passed `target.color` instead of `Radar::GetColor(target.type)`.

Comment (Hurleveur) [reply source/Engine.h:138]: oh right engine holds the target pointer, thx MZ
alr seems good

File: source/Radar.cpp [hunk 13]
```diff
@@ -143,8 +143,8 @@ const Color &Radar::GetColor(int type)
 		*GameData::Colors().Get("radar hostile"),
 		*GameData::Colors().Get("radar inactive"),
 		*GameData::Colors().Get("radar special"),
-		*GameData::Colors().Get("radar anomalous"),
 		*GameData::Colors().Get("radar blink"),
+		*GameData::Colors().Get("radar anomalous"),
```
Comment (quyykk) [inline source/Radar.cpp:147]: Accidental change you forgot to revert?

Comment (warp-core) [inline source/Radar.cpp:147]:
```suggestion
		*GameData::Colors().Get("radar anomalous"),
		*GameData::Colors().Get("radar blink"),
```

File: source/MapPlanetCard.cpp [hunk 14]
```diff
@@ -44,10 +44,26 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
+				break;
+			case Planet::Friendliness::RESTRICTED:
+				reputationLabel = "Restricted";
+				break;
+			case Planet::Friendliness::HOSTILE:
+				reputationLabel = "Hostile";
+				break;
+			case Planet::Friendliness::DOMINATED:
+				reputationLabel = "Dominated";
+				break;
```
Comment (warp-core) [inline source/MapPlanetCard.cpp:64]:
```suggestion
			case Planet::Friendliness::FRIENDLY:
				reputationLabel = "Friendly";
				break;
			case Planet::Friendliness::RESTRICTED:
				reputationLabel = "Restricted";
				break;
			case Planet::Friendliness::HOSTILE:
				reputationLabel = "Hostile";
				break;
			case Planet::Friendliness::DOMINATED:
				reputationLabel = "Dominated";
				break;
			default:
				reputationLabel = "Friendly";
				break;
```

File: source/MapPlanetCard.cpp [hunk 15]
```diff
@@ -44,10 +44,29 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
+			default:
+				reputationLabel = "Friendly";
+				break;
```
Comment (Hurleveur) [inline source/MapPlanetCard.cpp:67]:
```suggestion
			case Planet::Friendliness::RESTRICTED:
				reputationLabel = "Restricted";
				break;
			case Planet::Friendliness::HOSTILE:
				reputationLabel = "Hostile";
				break;
			case Planet::Friendliness::DOMINATED:
				reputationLabel = "Dominated";
				break;
			case Planet::Friendliness::FRIENDLY:
			default:
				reputationLabel = "Friendly";
				break;
```
you could just make it fall in like this

Comment (warp-core) [reply source/MapPlanetCard.cpp:67]:
```suggestion
			default:
			case Planet::Friendliness::FRIENDLY:
				reputationLabel = "Friendly";
				break;
			case Planet::Friendliness::RESTRICTED:
				reputationLabel = "Restricted";
				break;
			case Planet::Friendliness::HOSTILE:
				reputationLabel = "Hostile";
				break;
			case Planet::Friendliness::DOMINATED:
				reputationLabel = "Dominated";
				break;
```

Comment (Hurleveur) [reply source/MapPlanetCard.cpp]: Idk if that works like that but mb it does, ~~just make sure not everything is counted as Friendly~~
I've always seen the default below, but not sure if its needed

File: source/MapPlanetCard.cpp [hunk 16]
```diff
@@ -44,10 +44,27 @@ MapPlanetCard::MapPlanetCard(const StellarObject &object, unsigned number, bool
...
-		GameData::GetPolitics().HasDominated(planet) ? "Dominated" :
-		planet->GetGovernment()->IsEnemy() ? "Hostile" :
-		planet->CanLand() ? "Friendly" : "Restricted";
+	if(!hasSpaceport)
+		reputationLabel = "No Spaceport";
+	else
+	{
+		switch(planet->GetFriendliness())
+		{
+			default:
```
Comment (quyykk) [inline source/MapPlanetCard.cpp:53]: I would remove it tbh, because some compilers warn if an enum switch doesn't have a case for every value.

File: source/Planet.cpp [hunk 17]
```diff
@@ -557,6 +557,20 @@ bool Planet::CanLand() const
...
 
 
+Planet::Friendliness Planet::GetFriendliness() const
+{
+	if(GameData::GetPolitics().HasDominated(this))
+		return Friendliness::DOMINATED;
+	else if(CanLand())
+		return Friendliness::FRIENDLY;
+	else if(GetGovernment()->IsEnemy())
+		return Friendliness::HOSTILE;
```
Comment (quyykk) [inline source/Planet.cpp:567]: Switch these two around so that it's like how it originally was.

**Review by quyykk — APPROVED** [review]
